### DISCUSSION TITLE
Select does not show my selected value

### DISCUSSION THREAD

**Discussion by shahryarjb** [shadcn-ui/ui#1998] [Q&A]
Hello, please see these fields,  as you see the select of `shadcn` allways shows the **placeholder** `<SelectValue placeholder="انتخاب کنید" />`, and can not able to show default value of my api, but the normal html select show the value.

I have read all the topics the other suggest the value of `SelectItem` should be string, and my value is string too

![Screenshot 2023-11-23 at 10 44 50 in the morning](https://github.com/shadcn-ui/ui/assets/8413604/72cafc3a-18f0-4f58-a96f-30f5154c83c6)

my code:
```javascript
return (
    <div className={`w-full ${configs?.options?.classes}`}>
      <FormField
        control={form.control}
        name={configs.options.fieldName}
        render={({ field }) => (
          <FormItem>
            {configs?.name ? (
              <FormLabel className="text-black mb-2">{configs.name}</FormLabel>
            ) : null}
            <Select
              onValueChange={field.onChange}
              defaultValue={field.value}
            >
              <FormControl>
                <SelectTrigger>
                  <SelectValue placeholder="انتخاب کنید" />
                </SelectTrigger>
              </FormControl>
              <SelectContent>
                {Object.entries(options).map(([key, value]) => (
                      <SelectItem key={key} value={key}>
                        <>
                          {console.log(field.value, key)}
                          {value as React.ReactNode}
                        </>
                      </SelectItem>
                    ))
                  : null}
              </SelectContent>
            </Select>
            <FormMessage />
          </FormItem>
        )}
      />
    </div>
  );
```

I have changed the `defaultValue` to ` value` and etc. 

By the way, some times I changed an element in page without refresh `shadcn` select works, but if I refresh this, again you can not be able to see selected value.

I am using `nextjs` last version with app router and the last version of `react-hook-form` and `zod`

Thank you please help me to fix this issue

**Comment by jiaulislam** [comment]
> Hello, please see these fields,  as you see the select of `shadcn` allways shows the **placeholder** `<SelectValue placeholder="انتخاب کنید" />`, and can not able to show default value of my api, but the normal html select show the value.
> 
> I have read all the topics the other suggest the value of `SelectItem` should be string, and my value is string too
> 
> ![Screenshot 2023-11-23 at 10 44 50 in the morning](https://github.com/shadcn-ui/ui/assets/8413604/72cafc3a-18f0-4f58-a96f-30f5154c83c6)
> 
> my code:
> ```javascript
> return (
>     <div className={`w-full ${configs?.options?.classes}`}>
>       <FormField
>         control={form.control}
>         name={configs.options.fieldName}
>         render={({ field }) => (
>           <FormItem>
>             {configs?.name ? (
>               <FormLabel className="text-black mb-2">{configs.name}</FormLabel>
>             ) : null}
>             <Select
>               onValueChange={field.onChange}
>               defaultValue={field.value}
>             >
>               <FormControl>
>                 <SelectTrigger>
>                   <SelectValue placeholder="انتخاب کنید" />
>                 </SelectTrigger>
>               </FormControl>
>               <SelectContent>
>                 {Object.entries(options).map(([key, value]) => (
>                       <SelectItem key={key} value={key}>
>                         <>
>                           {console.log(field.value, key)}
>                           {value as React.ReactNode}
>                         </>
>                       </SelectItem>
>                     ))
>                   : null}
>               </SelectContent>
>             </Select>
>             <FormMessage />
>           </FormItem>
>         )}
>       />
>     </div>
>   );
> ```
> 
> I have changed the `defaultValue` to ` value` and etc. 
> 
> By the way, some times I changed an element in page without refresh `shadcn` select works, but if I refresh this, again you can not be able to see selected value.
> 
> I am using `nextjs` last version with app router and the last version of `react-hook-form` and `zod`
> 
> Thank you please help me to fix this issue

@shahryarjb did you able to solve the issue ? Facing same !

**Reply by shahryarjb** [reply]
@jiaulislam, I thank you I am not using shadcn anymore and replace it with another pkg. but thank you

**Reply by mluev** [reply]
you literally changed the whole ui kit because of this bro)?

**Comment by PratikAwaik** [comment]
@shadcn Facing this as well. Suddenly, in my app, all the Select dropdown stopped showing value in SelectTrigger. It just shows blank

**Reply by lanaschuster** [reply]
I'm also having the same problem. The alternative I found was to replace the SelectValue with another component, in my case a span.

```
<FormField
    control={form.control}
    name="status"
    render={({ field }) => (
        <FormItem>
            <FormLabel>Status *</FormLabel>
            <Select onValueChange={field.onChange} defaultValue={field.value}>
                <FormControl>
                    <SelectTrigger>
                        <span className="capitalize">
                            { field.value }
                        </span>
                    </SelectTrigger>
                </FormControl>
                <SelectContent>
                    <SelectItem value="ativado">Ativado</SelectItem>
                    <SelectItem value="desativado">Desativado</SelectItem>
                </SelectContent>
            </Select>
            <FormMessage />
        </FormItem>
    )}
/>
```

**Reply by PratikAwaik** [reply]
@lanaschuster yeah had a similar workaround!

**Comment by dMaisnam** [comment]
Since you were also using `react-hook-form` and `zod`, I managed to come across a similar issue where changing the options i.e `<SelectItem>` was not updating in the UI and during form submission, the default value keeps getting submitted. I managed to solve this by using the `<Controller />` component in `react-hook-form`
```tsx
const albumSchema = z.object({
    album_id: z.string().min(1, "Album is required"),
});

const { control } = useForm<IAlbum>({
    defaultValues: { album_id: "" },
    resolver: zodResolver(albumSchema)
});
```
```tsx
<Controller
    name="album_id"
    control={control}
    render={({ field }) => (
        <Select value={field.value} onValueChange={field.onChange}>
            <SelectTrigger>
                <SelectValue placeholder="Select Album" />
            </SelectTrigger>
            <SelectContent>
                {data.map(item => (
                    <SelectItem key={item.id} value={String(item.id)}>
                        {item.title}
                    </SelectItem>
                ))}
            </SelectContent>
        </Select>
    )}
/>
```

**Comment by jhovadev** [comment]
Yoo Jhoan here so i have the same issue but it was actually bc my data like:
code: number,
value: string 
so:
i look closely into my code and change the value from number to string into the value attr in selectitem and walah idk if this is right or helps someone pls answer me if this is correct or a mal practice ok 
`	<Select
															onValueChange={field.onChange}
															defaultValue={field.value}
														>
															<FormControl>
																<SelectTrigger>
																	<SelectValue placeholder="here pls" />{" "}
																</SelectTrigger>
															</FormControl>
															<SelectContent>
																{documentTypes.map((item: any) => (
																	<SelectItem
																		key={`${item.code}`}
																		value={`${item.code}`}
																	>
																		{item.value}
																	</SelectItem>
																))}
															</SelectContent>
														</Select>`

**Comment by smillwith** [comment]
Still testing this, but this looks like it works. Binding to a variable rather than the 'field.value' seems to reliably show the selected value and respond to change.

Before:

```typescript
<FormField
      control={form.control}
      name="homeLocationClientId"
      render={({ field }) => (
        <FormItem>
          <FormLabel>
            Home Location
            <RequiredIndicator />
          </FormLabel>
          <FormControl>
            <Select
              defaultValue={field.value || ""}
              onValueChange={field.onChange}
            >
```

After:

```typescript
const locationId = user?.homeLocationClientId ?? "";
...
<FormField
      control={form.control}
      name="homeLocationClientId"
      render={({ field }) => (
        <FormItem>
          <FormLabel>
            Home Location
            <RequiredIndicator />
          </FormLabel>
          <FormControl>
            <Select
              defaultValue={locationId || ""}
              onValueChange={field.onChange}
            >
```

**Comment by mluev** [comment]
> Hello, please see these fields, as you see the select of `shadcn` allways shows the **placeholder** `<SelectValue placeholder="انتخاب کنید" />`, and can not able to show default value of my api, but the normal html select show the value.
> 
> I have read all the topics the other suggest the value of `SelectItem` should be string, and my value is string too
> 
> ![Screenshot 2023-11-23 at 10 44 50 in the morning](https://private-user-images.githubusercontent.com/8413604/285122063-72cafc3a-18f0-4f58-a96f-30f5154c83c6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDk4MTIxMTIsIm5iZiI6MTc0OTgxMTgxMiwicGF0aCI6Ii84NDEzNjA0LzI4NTEyMjA2My03MmNhZmMzYS0xOGYwLTRmNTgtYTk2Zi0zMGY1MTU0YzgzYzYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MTNUMTA1MDEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmY3NmQwMTdiYjZkODIwZmYxMGJkOTFlNDk4MWEyMTBlMTlkZDhhNzU0ZGE0YTJiYmVjZjM5MWVlNWI5OTVhNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.T-R2GQhsymzCId29ieMZcyvnVEyvoehn4r5YIRviDkI)
> 
> my code:
> 
> ```js
> return (
>     <div className={`w-full ${configs?.options?.classes}`}>
>       <FormField
>         control={form.control}
>         name={configs.options.fieldName}
>         render={({ field }) => (
>           <FormItem>
>             {configs?.name ? (
>               <FormLabel className="text-black mb-2">{configs.name}</FormLabel>
>             ) : null}
>             <Select
>               onValueChange={field.onChange}
>               defaultValue={field.value}
>             >
>               <FormControl>
>                 <SelectTrigger>
>                   <SelectValue placeholder="انتخاب کنید" />
>                 </SelectTrigger>
>               </FormControl>
>               <SelectContent>
>                 {Object.entries(options).map(([key, value]) => (
>                       <SelectItem key={key} value={key}>
>                         <>
>                           {console.log(field.value, key)}
>                           {value as React.ReactNode}
>                         </>
>                       </SelectItem>
>                     ))
>                   : null}
>               </SelectContent>
>             </Select>
>             <FormMessage />
>           </FormItem>
>         )}
>       />
>     </div>
>   );
> ```
> 
> I have changed the `defaultValue` to ` value` and etc.
> 
> By the way, some times I changed an element in page without refresh `shadcn` select works, but if I refresh this, again you can not be able to see selected value.
> 
> I am using `nextjs` last version with app router and the last version of `react-hook-form` and `zod`
> 
> Thank you please help me to fix this issue

I had the same issue and i found the solution. 
My problem was that i get list with id: number from server and the selected value is probably string, so make sure that all values are string.

**Reply by sadamkhan7679** [reply]
Had the same issue...

It seems passing  items={options} to Select component fixes the issue

```
<Controller
      name={name}
      control={control}
      render={({ field, fieldState }) => {
        const isInvalid = fieldState.invalid;
        return (
          <Field data-invalid={isInvalid} className={containerClassName}>
            <FieldLabel aria-required={required} htmlFor={fieldId}>
              {label}
            </FieldLabel>
            <Select
              onValueChange={field.onChange}
              value={field.value || ""}
              items={options}
              {...rest}
            >
              <SelectTrigger aria-placeholder={placeholder} id={fieldId}>
                <SelectValue aria-placeholder={placeholder} />
              </SelectTrigger>
              <SelectContent>
                {options?.map((option) => (
                  <SelectItem key={option.value} value={option.value?.toString()}>
                    {option.label}
                  </SelectItem>
                ))}
              </SelectContent>
            </Select>
            {isInvalid && <FieldError errors={[fieldState.error]} />}
          </Field>
        );
      }}
    /> 
```
